src/Core/Services/SendService.cs

```diff
@@ -0,0 +1,275 @@
+            return sends.Select(kvp => new Send(kvp.Value)).ToList();
+        }
+
+        public async Task<List<SendView>> GetAllDecryptedAsync()
+        {
+            if (_decryptedSendsCache != null)
+            {
+                return _decryptedSendsCache;
+            }
+
+            var hasKey = await _cryptoService.HasKeyAsync();
+            if (!hasKey)
+            {
+                throw new Exception("No Key.");
+            }
+
+            if (_getAllDecryptedTask != null && !_getAllDecryptedTask.IsCompleted && !_getAllDecryptedTask.IsFaulted)
+            {
+                return await _getAllDecryptedTask;
+            }
+
+            async Task<List<SendView>> doTask()
+            {
+                var decSends = new List<SendView>();
+
+                async Task decryptAndAddSendAsync(Send send) => decSends.Add(await send.DecryptAsync());
+                await Task.WhenAll((await GetAllAsync()).Select(s => decryptAndAddSendAsync(s)));
+
+                decSends.OrderBy(s => s, new SendLocaleComparer(_i18nService)).ToList();
+                _decryptedSendsCache = decSends;
+                return _decryptedSendsCache;
+            }
+
+            _getAllDecryptedTask = doTask();
+            return await _getAllDecryptedTask;
+        }
+
+        public async Task<Send> GetAsync(string id)
+        {
+            var userId = await _userService.GetUserIdAsync();
+            var sends = await _storageService.GetAsync<Dictionary<string, SendData>>(SEND_KEY(userId));
+
+            if (sends == null || !sends.ContainsKey(id))
+            {
+                return null;
+            }
+
+            return new Send(sends[id]);
+        }
+
+        public async Task ReplaceAsync(Dictionary<string, SendData> sends)
+        {
+            var userId = await _userService.GetUserIdAsync();
+            await _storageService.SaveAsync(SEND_KEY(userId), sends);
+            _decryptedSendsCache = null;
+        }
+
+        public async Task SaveWithServerAsync(Send send, byte[] encryptedFileData)
+        {
+
+            var request = new SendRequest(send);
+            SendResponse response = null;
+            if (send.Id == null)
+            {
+                if (send.Type == SendType.Text)
+                {
+                    response = await _apiService.PostSendAsync(request);
+                }
+                else if (send.Type == SendType.File)
+                {
+                    var fd = new MultipartFormDataContent($"--BWMobileFormBoundary{DateTime.UtcNow.Ticks}")
+                    {
+                        { new StringContent(JsonConvert.SerializeObject(request)), "model" },
+                        { new ByteArrayContent(encryptedFileData), "data", send.File.FileName.EncryptedString }
+                    };
+
+                    response = await _apiService.PostSendFileAsync(fd);
+                }
+                send.Id = response?.Id;
```

Comment:
This line handles a `null` `response`, however what happens in this scenario upstream perhaps (is it possible?)

Reply:
Nothing else ever accesses this `response`.

To answer your question though, apiService will never return null for any of these calls (because `hasResponse` is set for all of them)

I null checked here out of habit due to assigning response null earlier.

Reply:
Actually looking again, this is a catch against unknown Send Types. I'm going to convert to a switch and make that explicit, and create an empty `SendResponse` if it happens